File: dispatch/java/adamma/c4dhi/claid/LocalDispatching/ModuleDispatcher.java
package adamma.c4dhi.claid.LocalDispatching;


import adamma.c4dhi.claid.ClaidServiceGrpc;
import adamma.c4dhi.claid.ClaidServiceGrpc.ClaidServiceStub;
import adamma.c4dhi.claid.Logger.Logger;
import adamma.c4dhi.claid.Module.ModuleFactory;

import adamma.c4dhi.claid.DataPackage;
import adamma.c4dhi.claid.ControlPackage;
import adamma.c4dhi.claid.CtrlType;
import adamma.c4dhi.claid.InitRuntimeRequest;
import adamma.c4dhi.claid.Runtime;
import adamma.c4dhi.claid.ModuleListRequest;
import adamma.c4dhi.claid.ModuleListResponse;
import adamma.c4dhi.claid.ModuleListResponse.ModuleDescriptor;

import io.grpc.Channel;
import io.grpc.ManagedChannel;
import io.grpc.Grpc;
import io.grpc.InsecureChannelCredentials;
import io.grpc.stub.StreamObserver;
import io.grpc.Status;
import io.grpc.StatusRuntimeException;

import com.google.protobuf.Empty;

import java.util.ArrayList;
import java.util.Map;


// Corresponds to ClientDispatcher of the C++ core, which is ModuleDispatcher in dart.
public class ModuleDispatcher  
{
    Channel grpcChannel;
    private ClaidServiceStub stub;
    
    private String socketPath;

    // Stream for us to receive packages from the Middleware.
    StreamObserver<DataPackage> inStream;

    // Stream for us to send packages to the Middleware
    StreamObserver<DataPackage> outStream;


    public ModuleDispatcher(final String socketPath)
    {
        this.socketPath = socketPath;

        this.grpcChannel = Grpc.newChannelBuilder(this.socketPath, InsecureChannelCredentials.create()).build();
        this.stub = ClaidServiceGrpc.newStub(this.grpcChannel);
    }

    // Tells the Middleware Server which Modules are available by this Runtime.
    // The Middleware will resond with the List of Modules that it wants this Runtime to initialize.
    // In general: Once the Middleware has started, it loads a config file. From the config file, it knows
    // which Modules to instantiate, but it does not yet know which Runtime (specifically, the ModuleDispatchers of each Runtime) can load which Module.
    // Therefore, it relies on the Runtimes to tell the Middleware which Modules they can handle. They do so by calling the stub.GetModuleList() function,
    // which contains a ModuleListRequest, which contains a list of all Modules the Runtime can handle.
    // The Middleware then replies with a ModuleListResponse, which contains the Modules it wants this Runtime to load (i.e., a subset of the Modules that were
    // sent in the ModuleListRequest).
    public ModuleListResponse getModuleList(ArrayList<String> registeredModuleClasses)
    {
        ModuleListRequest request = 
            ModuleListRequest.newBuilder()
            .setRuntime(Runtime.RUNTIME_JAVA)
            .addAllSupportedModuleClasses(registeredModuleClasses)
            .build();
        
        // stub.getModuleList() -> calls getModuleList function of RemoteService in the MiddleWare via RPC.

        boolean resultReceived = false;
        ModuleListResponse response = null;

        SynchronizedStreamObserver<ModuleListResponse> responseObserver = new SynchronizedStreamObserver<>();
        
        Logger.logInfo("Java Runtime: Calling getModuleList(...)");
        stub.getModuleList(request, responseObserver);

        response = responseObserver.await();
 

        return response;
    }


    public boolean initRuntime(Map<String, DataPackage> channelExamplePackages)
    {
        InitRuntimeRequest.Builder initRuntimeRequest = InitRuntimeRequest.newBuilder();
        
        for(Map.Entry<String, DataPackage> entry : channelExamplePackages.entrySet())
        {
            InitRuntimeRequest.ModuleChannels.Builder moduleChannels = InitRuntimeRequest.ModuleChannels.newBuilder();
            moduleChannels.setModuleId(entry.getKey());
            moduleChannels.addChannelPackets(entry.getValue());
            initRuntimeRequest.addModules(moduleChannels.build());
        }   
        initRuntimeRequest.setRuntime(Runtime.RUNTIME_JAVA);

        boolean resultReceived = false;
        Empty response = null;

        SynchronizedStreamObserver<Empty> responseObserver = new SynchronizedStreamObserver<>();

        Logger.logInfo("Java Runtime: Calling initRuntime(...)");
        stub.initRuntime(initRuntimeRequest.build(), responseObserver);

        response = responseObserver.await();
    
        
        return true;
    }


    private DataPackage makeControlRuntimePing()
    {
        DataPackage.Builder builder = DataPackage.newBuilder();
        ControlPackage.Builder controlPackageBuilder = ControlPackage.newBuilder();
        controlPackageBuilder.setCtrlType(CtrlType.CTRL_RUNTIME_PING);
        controlPackageBuilder.setRuntime(Runtime.RUNTIME_JAVA);

        builder.setControlVal(controlPackageBuilder.build());
        return builder.build();
    }

    public StreamObserver<DataPackage> sendReceivePackages(StreamObserver<DataPackage> inStream)
    {
        this.inStream = inStream;
        this.outStream = stub.sendReceivePackages(this.inStream);
   
  
        DataPackage pingReq = makeControlRuntimePing();
        this.outStream.onNext(pingReq);

        return this.outStream;

        // if (!stream->Write(pingReq)) {
        //     claid::Logger::printfln("Failed sending ping package to server.");
        //     return false;
        // }

        // // Wait for the valid response ping
        // DataPackage pingResp;
        // if (!stream->Read(&pingResp)) {
        //     claid::Logger::printfln("Did not receive a ping package from server !");
        //     return false;
        // }

        // TODO: @Stephan, shouldn't this here be pingResp ?
        // @Patrick:
        // // Check whether the package read was a control package with the right type.
        // if (pingReq.control_val().ctrl_type() != CtrlType::CTRL_RUNTIME_PING) {
        //     return false;
        // }

        // // Start the threads to service the input/output queues.
        // writeThread = make_unique<thread>([this]() { processWriting(); });
        // readThread = make_unique<thread>([this]() { processReading(); });
        // return true;*/

    }


    


    public void closeOutputStream()
    {
        this.outStream.onCompleted();
        this.outStream = null;
    }


    boolean postPackage(DataPackage packet)
    {
        return false;
    }



    // private ClaidService stub;
}

File: dispatch/python/local_dispatching/module_dispatcher.py
from dispatch.proto.sensor_data_types_pb2 import *
from dispatch.proto.claidservice_pb2 import * 
from dispatch.proto.claidservice_pb2_grpc import ClaidServiceStub

from logger.logger import Logger

import grpc

from timeit import time

from module.thread_safe_channel import ThreadSafeChannel

class ModuleDispatcher:
    def __init__(self, socket_path):
        self.socket_path = socket_path
  
        self.grpc_channel = grpc.insecure_channel(socket_path)

        self.stub = ClaidServiceStub(self.grpc_channel)

        # self.to_middleware_queue = io.Queue()
        self.__to_middleware_queue = ThreadSafeChannel()
        self.__from_middleware_queue = None
        self.__running = False


    def get_to_dispatcher_queue(self):
        return self.__to_middleware_queue
    
    def get_from_dispatcher_queue(self):
        return self.__from_middleware_queue

    def get_module_list(self, registered_module_classes, module_annotations):

      

        request = ModuleListRequest(
            runtime=Runtime.RUNTIME_PYTHON,
            supported_module_classes=registered_module_classes,
            module_annotations=module_annotations
        )
        Logger.log_info(f"Sending ModuleListRequest: {request}")

      

        Logger.log_info("Python Runtime: Calling getModuleList(...)")

        try:
            response =  self.stub.GetModuleList(request)

            if response is None:
                Logger.log_error(f"Error occurred in getModuleList() of PYTHON_RUNTIME: {response_observer.get_error_message()}")
                exit(0)
        except Exception as e:
            print(e)
            return False

        print(f"Response: {response}")
        return response

    def init_runtime(self, channel_example_packages):
        init_runtime_request = InitRuntimeRequest()

        print(channel_example_packages)
        for module_id in channel_example_packages:
            channel_packages = channel_example_packages[module_id]
            module_channels = InitRuntimeRequest.ModuleChannels(
                module_id=module_id,
                channel_packets=channel_packages
            )
            init_runtime_request.modules.append(module_channels)

        init_runtime_request.runtime = Runtime.RUNTIME_PYTHON

    
        Logger.log_info("Python Runtime: Calling initRuntime(...)")
        response =  self.stub.InitRuntime(init_runtime_request)


        if response is None:
            Logger.log_error(f"Error occurred in initRuntime() of Python: respone of InitRuntime is None")
            return False

        return True

    def make_control_runtime_ping(self):
        package = DataPackage()
        control_package = ControlPackage(
            ctrl_type=CtrlType.CTRL_RUNTIME_PING,
            runtime=Runtime.RUNTIME_PYTHON
        )

        package.control_val.CopyFrom(control_package)
        return package

  
    def to_middleware_queue_get(self):

        while self.__running:
            data = self.__to_middleware_queue.get()
            if not data is None:
                yield data
            elif not self.__running:
                return


    def send_receive_packages(self):

        self.__running = True
        self.__from_middleware_queue = self.stub.SendReceivePackages(self.to_middleware_queue_get())

        if self.__to_middleware_queue is None:
            Logger.log_error("Failed to initialize streaming to/from middleware: stub.sendReceivePackages return value is None.")
            return False

        self.waiting_for_ping_response = True
        ping_req = self.make_control_runtime_ping()

        Logger.log_info("Putting package")
        self.__to_middleware_queue.put(ping_req)

        ping_resp = self.await_ping_package()

        if(ping_resp.control_val.ctrl_type != CtrlType.CTRL_RUNTIME_PING):
            Logger.log_error(f"Sent ping package to middleware and expected ping package as response, but got: {ping_resp}")
            return False

        return True

    def await_ping_package(self):
        Logger.log_info("Waiting for ping package")
        for response in self.__from_middleware_queue:
            Logger.log_info(f"Got ping package {response}")
            return response
        
    def shutdown(self):
        self.__running = False
        self.__to_middleware_queue.put(None)
       
    # def on_middleware_stream_package_received(self, packet):
    #     Logger.log_info(f"Java Runtime received message from middleware: {packet}")

    #     if self.waiting_for_ping_response:
    #         if packet.control_val.ctrl_type != CtrlType.CTRL_RUNTIME_PING:
    #             Logger.log_error("Error in ModuleDispatcher: Sent CTRL_RUNTIME_PING to middleware and was waiting for a package with the same type as response,\n"
    #                              f"but got package of type {packet.control_val}")
    #             return

    #         if packet.control_val.runtime != Runtime.RUNTIME_JAVA:
    #             Logger.log_error("Error in ModuleDispatcher: Sent CTRL_RUNTIME_PING to middleware and received response, however response has wrong Runtime identifier.\n"
    #                              f"Expected RUNTIME_JAVA but got {packet.control_val.runtime}\n")
    #             return

    #         Logger.log_info("Java ModuleDispatcher successfully registered at the Middleware!")

    #         self.waiting_for_ping_response = False
    #         return

    #     Logger.log_info(f"Received package {packet.control_val}")
    #     self.in_consumer(packet)

    # def on_middleware_stream_error(self, throwable):
    #     Logger.log_error(f"Middleware stream closed! {throwable}")
    #     self.close_output_stream()

    # def on_middleware_stream_completed(self):
    #     Logger.log_error("Middleware stream completed!")
    #     self.close_output_stream()

    # def close_output_stream(self):
    #     self.out_stream.on_completed()
    #     self.out_stream = None

    # def post_package(self, packet):
    #     self.out_stream.on_next(packet)
    #     return False
